File: coord_pixel_transfer_py/coord_pixel_transfer.py
#!/usr/bin/python
# coding: utf-8



from math import *



class CoordPixelTransfer:
    def __init__(self):
        self.r              = 128/pi
        self.rad            = pi/180
        self.zoom_level_min = 0
        self.zoom_level_max = 20
        self.tile_width     = 256
        self.tile_height    = 256
        self.lat_min        = float("%.8f" % float(atan(sinh(((128 - (256 -   0)) * pi) / 128)) * 180/pi))
        self.lat_max        = float("%.8f" % float(atan(sinh(((128 - (256 - 256)) * pi) / 128)) * 180/pi))
        self.lng_min        = float("%.8f" % float((((  0 * pi) / 128) - pi) * 180/pi))
        self.lng_max        = float("%.8f" % float((((256 * pi) / 128) - pi) * 180/pi))


    def decimalFormat(self, num):
        return float("%.8f" % num)


    def checkLatlngRange(self, latlng):
        lat, lng = self.decimalFormat(latlng[0]), self.decimalFormat(latlng[1])
        if (not self.lat_min < lat < self.lat_max) or (not self.lng_min < lng < self.lng_max): 
            raise ValueError("latlng out of range")


    def checkTilePixelRange(self, p):
        px, py = p[0], p[1]
        if (not 0 < px < self.tile_width) or (not 0 < py <  self.tile_height):
            raise ValueError("world coordinate out of range world")
        

    def checkZoomRange(self, zoom):
        if not self.zoom_level_min < zoom < self.zoom_level_max:
            raise ValueError("zoom out of range")


    def fromLatLngToWorldCoord(self, latlng):
        self.checkLatlngRange(latlng)

        lat, lng = self.decimalFormat(latlng[0]), self.decimalFormat(latlng[1])
        lat_rad, lng_rad = lat * self.rad, lng * self.rad

        return (
            self.decimalFormat( self.r * (lng_rad + pi) ),
            self.decimalFormat( -self.r/2 * log((1 + sin(lat_rad)) / (1 - sin(lat_rad))) + 128 ))

    def fromLatLngToWorldCoordRatio(self, latlng):
        x, y = self.fromLatLngToWorldCoord(latlng)

        return ( x / float(self.tile_width), y / float(self.tile_height) )


    def fromWorldCoordToLatLng(self, p):
        self.checkTilePixelRange(p)
        
        px, py = self.decimalFormat(p[0]), self.decimalFormat(p[1])

        return (
            self.decimalFormat( atan(sinh(((128 - py) * pi) / 128)) * 180/pi ),
            self.decimalFormat( (((px * pi) / 128) - pi) * 180/pi ))


    def fromLatLngToPixelCoord(self, latlng, zoom):
        self.checkZoomRange(zoom)

        px, py = self.fromLatLngToWorldCoord(latlng)

        return ( px * (2 ** zoom), py * (2** zoom) )


    def fromPixelCoordToLatLng(self, p, zoom):
        self.checkZoomRange(zoom)

        px, py = (
            self.decimalFormat(p[0] / (2 ** zoom)), 
            self.decimalFormat(p[1] / (2 ** zoom)))

        return self.fromWorldCoordToLatLng((px, py))


    def fromLatLngToTileCoord(self, latlng, zoom):
        px, py = self.fromLatLngToPixelCoord(latlng, zoom)

        return (
            (int("%u" % floor(px / self.tile_width)), int("%u" % floor(py / self.tile_height))),
            (self.decimalFormat(px % self.tile_width), self.decimalFormat(py % self.tile_height)))


    def fromTileCoordToLatLng(self, p, zoom):
        px, py = p[0] * self.tile_width, p[1] * self.tile_height

        return self.fromPixelCoordToLatLng((px, py), zoom)


    def zoomRange(self):
        return (self.zoom_level_min, self.zoom_level_max)


    def test(self):
        print "lat_min", self.lat_min
        print "lat_max", self.lat_max
        print "lng_min", self.lng_min
        print "lng_max", self.lng_max

        latlng = (34.1234, 134.1234)

        print self.fromWorldCoordToLatLng((2, 23))
        print self.fromLatLngToWorldCoord(latlng)
        print self.fromLatLngToPixelCoord(latlng, 7)
        print self.fromLatLngToTileCoord(latlng, 7)

File: coord_pixel_transfer_cpp/coord_pixel_transfer.cpp
// huhuhu

#include <cstdio>
#include <cstdarg>
#include <cmath>
#include <tuple>
#include <stdexcept>
#include <str2num.h>
#include "coord_pixel_transfer.h"


namespace hoshi
{


using namespace std;
using namespace boost;



coord_pixel_transfer::coord_pixel_transfer() :
    ZOOM_LEVEL_MIN(0), 
    ZOOM_LEVEL_MAX(19), 
    TILE_WIDTH(256), 
    TILE_HEIGHT(256), 
    // LAT_MIN(str2num<double>("%.16f", (atan(sinh(((128 - (256 -   0)) * M_PI) / 128))) * 180/M_PI)), 
    // LAT_MAX(str2num<double>("%.16f", (atan(sinh(((128 - (256 - 256)) * M_PI) / 128))) * 180/M_PI)), 
    // LNG_MIN(str2num<double>("%.16f", (((  0 * M_PI) / 128) - M_PI) * 180/M_PI)), 
    // LNG_MAX(str2num<double>("%.16f", (((256 * M_PI) / 128) - M_PI) * 180/M_PI)),
    LAT_MIN((atan(sinh(((128 - (256 -   0)) * M_PI) / 128))) * 180/M_PI), 
    LAT_MAX((atan(sinh(((128 - (256 - 256)) * M_PI) / 128))) * 180/M_PI), 
    LNG_MIN((((  0 * M_PI) / 128) - M_PI) * 180/M_PI), 
    LNG_MAX((((256 * M_PI) / 128) - M_PI) * 180/M_PI),
    PI(M_PI), 
    R(128/M_PI)
{
}


coord_pixel_transfer::~coord_pixel_transfer()
{
}


tuple<double, double> coord_pixel_transfer::fromLatLngToWorldCoord(double lat, double lng)
{
    // lat = str2num<double>("%.16f", lat);
    // lng = str2num<double>("%.16f", lng);

    if((lat < LAT_MIN || lat > LAT_MAX) || (lng < LNG_MIN || lng > LNG_MAX)) {
        throw new out_of_range("out of range latlng");
    }

    double lat_rad = lat * PI/180;
    double lng_rad = lng * PI/180;

    return make_tuple(
        // str2num<double>("%.16f", R * (lng_rad + PI)), 
        // str2num<double>("%.16f", -R/2 * log((1 + sin(lat_rad)) / (1 - sin(lat_rad))) + 128));
        R * (lng_rad + PI), -R/2 * log((1 + sin(lat_rad)) / (1 - sin(lat_rad))) + 128);
}


tuple<double, double> coord_pixel_transfer::fromWorldCoordToLatLng(double wx, double wy)
{
    if((wx < 0 || wx > TILE_WIDTH) || (wy < 0 || wy > TILE_HEIGHT)) {
        throw new out_of_range("out of range world coordinate");
    }

    return make_tuple(
        // str2num<double>("%.16f", (atan(sinh(((128 - wy) * PI) / 128))) * 180/PI), 
        // str2num<double>("%.16f", (((wx * PI) / 128) - PI) * 180/PI));
        (atan(sinh(((128 - wy) * PI) / 128))) * 180/PI, (((wx * PI) / 128) - PI) * 180/PI);

}

tuple<unsigned int, unsigned int> coord_pixel_transfer::fromLatLngToPixelCoord(double lat, double lng, double zoomLevel)
{
    if(zoomLevel < ZOOM_LEVEL_MIN || zoomLevel > ZOOM_LEVEL_MAX) {
        throw new out_of_range("out of range zoom level");
    }

    tuple<double, double> wc;

    try {
        wc = this->fromLatLngToWorldCoord(lat, lng);
    } catch(out_of_range&) {
        throw;
    }

    return make_tuple(
        floor(get<0>(wc) * pow(2, zoomLevel)),
        floor(get<1>(wc) * pow(2, zoomLevel)));
}


tuple<double, double> coord_pixel_transfer::fromPixelCoordToLatLng(unsigned int px, unsigned int py, double zoomLevel)
{
    if(zoomLevel < ZOOM_LEVEL_MIN || zoomLevel > ZOOM_LEVEL_MAX) {
        throw new out_of_range("out of range zoom level");
    }

    // double wx = str2num<double>("%.16f", px / pow(2, zoomLevel));
    // double wy = str2num<double>("%.16f", py / pow(2, zoomLevel));
    double wx = px / pow(2, zoomLevel);
    double wy = py / pow(2, zoomLevel);

    tuple<double, double> latlng;

    try {
        latlng = this->fromWorldCoordToLatLng(wx, wy);
    } catch(out_of_range) {
        throw;
    }

    return latlng;
}


tuple<double, double, unsigned int, unsigned int>
coord_pixel_transfer::fromLatLngToTileCoord(double lat, double lng, double zoomLevel)
{
    tuple<unsigned int, unsigned int> pc;

    try {
        pc = this->fromLatLngToPixelCoord(lat, lng, zoomLevel);
    } catch(out_of_range&) {
        throw;
    }

    return make_tuple(
        // str2num<double>("%u", floor(get<0>(pc) / TILE_WIDTH)), 
        // str2num<double>("%u", floor(get<1>(pc) / TILE_HEIGHT)),
        // // タイル内のピクセル座標
        // str2num<unsigned int>("%.16f", get<0>(pc) % TILE_WIDTH),
        // str2num<unsigned int>("%.16f", get<1>(pc) % TILE_HEIGHT));
        floor(get<0>(pc) / TILE_WIDTH), 
        floor(get<1>(pc) / TILE_HEIGHT),
        // タイル内のピクセル座標
        get<0>(pc) % TILE_WIDTH,
        get<1>(pc) % TILE_HEIGHT);
}


tuple<double, double> coord_pixel_transfer::fromTileCoordToLatLng(unsigned int tx, unsigned int ty, double zoomLevel)
{
    unsigned int px = tx * TILE_WIDTH;
    unsigned int py = tx * TILE_HEIGHT;

    tuple<double, double> latlng;

    try {
        latlng = fromPixelCoordToLatLng(px, py, zoomLevel);
    } catch(out_of_range&) {
        throw;
    }

    return latlng;
}


tuple<double, double> coord_pixel_transfer::zoomLevelRange()
{
    return make_tuple(ZOOM_LEVEL_MIN, ZOOM_LEVEL_MAX);
}


}

